src/overlays/actors/ovl_En_Bubble/z_en_bubble.c

```diff
@@ -32,73 +31,401 @@ const ActorInit En_Bubble_InitVars = {
+}
+
+u32 func_8089F8BC(EnBubble* this) {
+    if (DECR(this->explosionCountdown)) {
+        return -1;
+    } else {
+        return func_8089F5D0(this);
+    }
+}
+
+s32 func_8089F908(EnBubble* this) {
+    this->expansionWidth += 1.0f / 12.0f;
+    this->expansionHeight += 1.0f / 12.0f;
+
+    if (DECR(this->explosionCountdown) != 0) {
+        return false;
+    } else {
+        return true;
+    }
+}
+
+void EnBubble_Vec3fNormalizedReflect(Vec3f* vec1, Vec3f* vec2, Vec3f* ret) {
+    f32 norm;
+
+    func_80179F64(vec1, vec2, ret);
+    norm = sqrtf((ret->x * ret->x) + (ret->y * ret->y) + (ret->z * ret->z));
+    if (norm != 0.0f) {
+        ret->x /= norm;
+        ret->y /= norm;
+        ret->z /= norm;
+    } else {
+        ret->x = ret->y = ret->z = 0.0f;
+    }
+}
+
+void EnBubble_Vec3fNormalize(Vec3f* vec) {
+    f32 norm = sqrtf((vec->x * vec->x) + (vec->y * vec->y) + (vec->z * vec->z));
+
+    if (norm != 0.0f) {
+        vec->x /= norm;
+        vec->y /= norm;
+        vec->z /= norm;
+    } else {
+        vec->x = vec->y = vec->z = 0.0f;
+    }
+}
+
+void EnBubble_Fly(EnBubble* this, PlayState* play) {
+    CollisionPoly* poly;
+    Actor* bumpActor;
+    Vec3f sp84;
+    Vec3f sp78;
+    Vec3f sp6C;
+    Vec3f sp60;
+    Vec3f sp54;
+    f32 bounceSpeed;
+    s32 bgId;
+    u8 bounceCount;
+
+    if (this->colliderSphere.elements[1].info.bumperFlags & BUMP_HIT) {
+        bumpActor = this->colliderSphere.base.ac;
+        this->normalizedBumpVelocity = bumpActor->velocity;
+        EnBubble_Vec3fNormalize(&this->normalizedBumpVelocity);
+        this->velocityFromBump.x += (this->normalizedBumpVelocity.x * 3.0f);
+        this->velocityFromBump.y += (this->normalizedBumpVelocity.y * 3.0f);
+        this->velocityFromBump.z += (this->normalizedBumpVelocity.z * 3.0f);
+    }
+    this->sinkSpeed -= 0.1f;
+    if (this->sinkSpeed < this->actor.terminalVelocity) {
+        this->sinkSpeed = this->actor.terminalVelocity;
+    }
+    sp54.x = this->velocityFromBounce.x + this->velocityFromBump.x;
+    sp54.y = this->velocityFromBounce.y + this->velocityFromBump.y + this->sinkSpeed;
+    sp54.z = this->velocityFromBounce.z + this->velocityFromBump.z;
+    EnBubble_Vec3fNormalize(&sp54);
+
+    sp78.x = this->actor.world.pos.x;
+    sp78.y = this->actor.world.pos.y + this->actor.shape.yOffset;
+    sp78.z = this->actor.world.pos.z;
+    sp6C = sp78;
+
+    sp6C.x += (sp54.x * 24.0f);
+    sp6C.y += (sp54.y * 24.0f);
+    sp6C.z += (sp54.z * 24.0f);
+    if (BgCheck_EntityLineTest1(&play->colCtx, &sp78, &sp6C, &sp84, &poly, true, true, true, false, &bgId)) {
+        sp60.x = COLPOLY_GET_NORMAL(poly->normal.x);
+        sp60.y = COLPOLY_GET_NORMAL(poly->normal.y);
+        sp60.z = COLPOLY_GET_NORMAL(poly->normal.z);
+        EnBubble_Vec3fNormalizedReflect(&sp54, &sp60, &sp54);
+        this->bounceDirection = sp54;
+        bounceCount = this->bounceCount;
+        this->bounceCount = ++bounceCount;
```

Comment:
Guessing it has to be like this?

Reply:
unfortunately, yes. It doesn't look very nice